### PR TITLE
Work with arduinostl

### PR DESCRIPTION
**Pull request by matthijskooijman**
See #1

### CODE REVIEW

File: ArxTypeTraits/type_traits.h [hunk 1]
```diff
@@ -378,15 +389,12 @@ namespace std {
...
-    }
-
+} // namespace std
 
+#if __has_include(<initializer_list>)
+#include <initializer_list>
+#else
+// ArxContainer might also define initializer_list, so don't redefine
 #ifndef ARX_TYPE_TRAITS_INITIALIZER_LIST_DEFINED
 #define ARX_TYPE_TRAITS_INITIALIZER_LIST_DEFINED
```
Comment (hideakitai) [inline ArxTypeTraits/type_traits.h]: Please add `std` namespace to `initializer_list`

**Comment by hideakitai** [comment]
And at the right time, can you add example which use this library with ArduinoSTL?

File: ArxTypeTraits/type_traits.h (hunk 1 above)
Comment (matthijskooijman) [reply ArxTypeTraits/type_traits.h]: Good point.

**Comment by matthijskooijman** [comment]
I forgot to post my test sketch yesterday, but please find it below now. I'm not sure whether this would be good for an actual example, though, as it just tries to cover some corner cases which would be weird in a real-world example.

```C++
#if 1 // Either include ArxTypeTraits or standard system headers (setting this to 0 might break e.g. void_t below)
#include "ArxTypeTraits.h"
#else
#include <functional>
#include <limits>
#endif

#if 0 // Check with standard libray or just ArxTypeTraits
#if __AVR__
#include <ArduinoSTL.h>
#endif
#include <limits>
#else
// Allow using arx::std using the std:: name
//namespace std { using namespace arx::arx_std; }
#endif

// ArxTypeTraits handles these, but undef these in case we're not including ArxTypeTraits
#undef min
#undef max

// This breaks currently
//using namespace arx;

std::function<void()> foo = []() {};

// When using ArduinoSTL, this is *not* constexpr
int y = std::numeric_limits<int>::max();
// But this is :-D
constexpr int z = arx::arx_std::numeric_limits<int>::max();

// Check that void_t is available and not ambiguous
using x = std::void_t<int>;

void setup() {
  Serial.begin(115200);
  randomSeed(analogRead(0));
  if (random(2)) {
    foo = []() {
      Serial.println("A");
    };
  } else {
    new (&foo) std::function<void()>([]() {
      Serial.println("B");
    });
  }
}

void loop() {
  // put your main code here, to run repeatedly:
  foo();
}
```

**Review by hideakitai — APPROVED** [review]

**Comment by matthijskooijman** [comment]
Cool! I'll try to clean up the history a bit more before merging this, though.

I'm also currently trying to compile this under cygwin, just to doublecheck whether that works.

**Comment by hideakitai** [comment]
Ok, thanks! I'll check it again after the refactor :)

I think ArxTypeTraits itself looks perfect, but compiling your example with ArduinoSTL (making this flag to 1) causes some errors.

```
#if 1 // Check with standard libray or just ArxTypeTraits
```

```
ArduinoSTL/src/del_opvs.cpp:25:53: error: 'std::size_t' has not been declared
 _UCXXEXPORT void operator delete[](void * ptr, std::size_t) throw(){

ArduinoSTL/src/del_ops.cpp:25:50: error: 'std::size_t' has not been declared
 _UCXXEXPORT void operator delete(void* ptr, std::size_t) throw(){

ArduinoSTL/src/del_opnt.cpp:25:56: error: 'nothrow_t' in namespace 'std' does not name a type
 _UCXXEXPORT void operator delete(void* ptr, const std::nothrow_t& ) throw() {

ArduinoSTL/src/del_opvnt.cpp:25:58: error: 'nothrow_t' in namespace 'std' does not name a type
 _UCXXEXPORT void operator delete[](void* ptr, const std::nothrow_t& ) throw(){
```

Though I have tried it with [original one](https://github.com/mike-matera/ArduinoSTL), do you use one of forks?

**Comment by matthijskooijman** [comment]
Oh, right, provided you're testing on AVR, that's this issue: https://github.com/mike-matera/ArduinoSTL/issues/56

It happens regardless of ArxTypeTraits, it's a conflict due to the Arduino core offering the `<new>` header since recently on AVR. A workaround is to downgrade the AVR core to 1.8.2, which is what I've been testing with so far. This should eventually be fixed in Arduino and/or ArduinoSTL, I think (I already submitted a PR for making `<new>` complete in Arduino, and have done some work there).

**Comment by hideakitai** [comment]
Ah, it's cool, thanks!

File: ArxTypeTraits/type_traits.h [hunk 2]
```diff
@@ -378,18 +388,17 @@ namespace std {
...
+// Initializer_list *must* be defined in std, so take extra care to only
+// define it when <initializer_list> is really not available (e.g.
+// ArduinoSTL is C++98 but *does* define <initializer_list>) and not
+// already defined (e.g. by ArxContainer).
+#if __has_include(<initializer_list>)
+#include <initializer_list>
+#elif !defined(ARX_TYPE_TRAITS_INITIALIZER_LIST_DEFINED)
 #define ARX_TYPE_TRAITS_INITIALIZER_LIST_DEFINED
-
+namespace arx_std {
```
Comment (hideakitai) [inline ArxTypeTraits/type_traits.h]: This should be `std` instead of `arx_std`?

Comment (matthijskooijman) [reply ArxTypeTraits/type_traits.h]: Yup, too eager search-replace :-) Pushed a fix just now.

**Comment by matthijskooijman** [comment]
Ok, a couple of dozen rebases later and I think the history looks nice and clean now. I now did the namespace rename before the library-detection refactor, and it turns out the namespace rename was actually enough to make things work with ArduinoSTDL (but the refactor is still nice, because the code is now cleaner and only defines what is really needed). I also split off some smaller commits here and there and improved the commit messages, should be easy to review now.

I also moved the `initializer_list` into its own file as discussed and did some other small improvements.

I think that this PR is ready for another review and merging as far as I'm concerned :-)

**Comment by matthijskooijman** [comment]
Here's a slightly updated version of the example, which also tests `initializer_list`: 
```C++
#include "ArxTypeTraits.h"

#if 0// Check with standard libray or just ArxTypeTraits
#if __AVR__
#include <ArduinoSTL.h>
#endif
#include <limits>
#endif

// Check this doesn't break
using namespace arx;

std::function<void()> foo = []() {};

// When using ArduinoSTL, this is *not* constexpr
int y = std::numeric_limits<int>::max();
// But this is :-D
constexpr int z = arx::arx_std::numeric_limits<int>::max();

// Check that void_t is available and not ambiguous
using x = std::void_t<int>;

// Check that initializer_list works
size_t listlen(std::initializer_list<int> s) { return s.size(); }
const size_t len = listlen({1, 2, 3});

void setup() {
  Serial.begin(115200);
  randomSeed(analogRead(0));
  if (random(2)) {
    foo = []() {
      Serial.println("A");
    };
  } else {
    new (&foo) std::function<void()>([]() {
      Serial.println("B");
    });
  }
}

void loop() {
  // put your main code here, to run repeatedly:
  foo();
}

```

 I compile-tested this on ESP32, SAMD and AVR.

**Comment by hideakitai** [comment]
Thanks so much! I've checked this PR also in my several libraries which uses ArxTypeTraits. Currently only one problem is about ESP8266. ESP8266's `__has_include` seems not working properly, it causes compile errors. Adding `|| defined(ESP8266)` to `__has_include` flags like `#if __has_include(<cstdlib>) || defined(ESP8266)` works fine, but... not smart is it? What do you think about it?

**Comment by matthijskooijman** [comment]
Hm, it seems that ESP8266 still uses gcc 4.8, and `__has_include` was introduced in gcc 5.0: https://gcc.gnu.org/gcc-5/changes.html

However, in my local test, I get errors about redefinition of `initializer_list`, as if `__has_include` just returned false, rather than this error: https://github.com/hideakitai/ArxTypeTraits/blob/eba2f99995aa744c76a5b13ccc438bf49c3614de/ArxTypeTraits.h#L11-L13

Turns out the ESP8266 libc defines this macro manually: https://github.com/esp8266/Arduino/blob/cc042b99d1c966a8b3e92ca9b7ffb3b497d0af98/tools/sdk/libc/xtensa-lx106-elf/include/sys/cdefs.h#L79-L81

Which is a bit of a bummer, since to ArxTypeTraits work on ESP8266, `__has_include` just needs to return true for all the checks we do, but we can't just redefine it. I guess we could define our own `ARX_HAS_INCLUDE(x)` which then just forwards to `__has_include` except on ESP8266, where it just returns true? I'll give that a try.

**Comment by hideakitai** [comment]
Thanks, it looks good!

> However, in my local test, I get errors about redefinition of initializer_list, as if __has_include just returned false, rather than this error:

Ah yes, I also got about initializer_list and placement new.

**Comment by matthijskooijman** [comment]
Done, see the latest commit. Compiletest on ESP8266, ESP32, SAMD and AVR (the last one with and without ArduinoSTL).

**Review by hideakitai — APPROVED** [review]
All commits looks fine and clear, and testing on my libraries depending on ArxTypeTraits also works fine. Thank you so much for your awesome contribution to ArxTypeTraits! 👏

**Comment by hideakitai** [comment]
After merging, I made some modifications I missed.
- move ARX_HAVE_LIBSTDCPLUSPLUS to separated file for portability (to another Arx Series)
- DUE has compile error because compiler is GCC 4.8.3 and no stdlibc++, so force `ARX_HAVE_LIBSTDCPLUSPLUS 0`
- fix `std::move` ambiguity (this is long existed bug which is not related to this PR)

Please confirm them and let me know if you have any comments :)

**Comment by matthijskooijman** [comment]
I left some inline comments on some of your issues.

As for Due, your observation surprises me, since (just checked), it *does* have libstdc++. But when I use the same approach as for ESP8266 (let `__has_include` return 1), then I get:
```
/tmp/arduino_build_static/sketch/test.cpp.o: In function `loop':
/tmp/arduino_build_static/sketch/test.cpp:47: undefined reference to `std::__throw_bad_function_call()'
collect2: error: ld returned 1 exit status
```

So that's a bit of a different problem. In fact, the problem seems to be that linking is done by `gcc` rather than `g++`. This should probably be fixed in the SAM core, but for now just ignoring the libstdc++ on Due is maybe not too bad (can always revert this when SAM is fixed). I also think SAM should (and can) be upgrade to a new gcc (same as SAMD), so I'll see if I can nudge people to do that as well :-)

**Comment by hideakitai** [comment]
Thanks for your comment, I've fixed those comments.

And sorry when I thought `!defined(__has_include)` line means *does not have libstdc++* (of course I know it's not...) even though it is arm board... I will do things more carefully :( Anyway, this would be better to 1 not 0, but currently remain as-is because of the error caused by std::function you mentioned. I will wait for you to nudge people :)

Thanks again!

**Comment by hideakitai** [comment]
If you can, can I ask you how you find which is the problem inside GCC (g++) like this?

> the problem seems to be that linking is done by gcc rather than g++

I'm not confident about such details currently, but I want to be go deeper if I can. I could go to functional of arm-none-eabi 4.8.3 inside of sam directory, but I couldn't go further like you. How could you find that cause? If you can, let me know. (The only way is to read and understand codes inside such toolchains maybe...)

Thanks, in advance.

**Comment by matthijskooijman** [comment]
> How could you find that cause?

I first switched SAM from gcc 4 to gcc 7 (the one from SAMD). I used a bit of a hack, replacing the 4.8.something directory in `~/.arduino/packages/arduino/tools/arm-gcc-none-eabi/` (or thereabouts) with a symlink to the 7 version. After fixing a problem with `printf`, I got this linker error. Iooked to see where this `std::__throw_bad_function_call` was supposed to be defined (grepping around the SAM and SAMD cores and both gcc versions) but couldn't find any place. Because it works for SAMD, I knew it had to come from *somewhere*. So I compared compiler commands between SAM and SAMD, then noticed gcc vs g++ and realized that g++ just magically inserts these symbols apparently. I left out some details and other small things I tried, but this was the general flow :-)

**Comment by hideakitai** [comment]
Thank you, great. I should go more along with the compiler... This library was the result of my research to understand what the type_traits works (and how to write it), so I will go further with compiler by improving (and debugging) this library :) Thank you so much! 
If you have nothing to comment currently, I will release the next version.

**Comment by matthijskooijman** [comment]
For the Due, I filed a fix for the linker error [here](https://github.com/arduino/ArduinoCore-sam/pull/111) and suggested upgrading the compiler [here](https://github.com/arduino/ArduinoCore-sam/issues/112).

Some more minor things:
 - #3 
 - The README now says "C++11 (defined only for platforms above which cannot use type_traits)", but there's no platforms defined above anymore. Maybe put this under "Supported Class Templates" instead?
 > When something is available from the standard library, it is included and can be used directly. If something from the below list is missing from the standard library, it is defined by ArxTypeTraits and can be accessed through the `std` namespace as normal. When the standard library defines an older version of a name (e.g. the non-constexpr C++98 version of `numeric_limits` defined by ArduinoSTL), then Arx will define the newer version (e.g. the constexpr C++11 version of `numeric_limits`), but it must then be accessed through the `arx::arx_std` namespace (since the standard library is preferred when using the `std` namespace).
 - I still wonder how nice `arx::arx_std` really is. I wonder if we shouldn't rename to e.g. `arx::stdx` (as you suggested somewhat earlier) instead, which looks cleaner to me. Should be still ok to rename now without any compatibility issues.

**Comment by hideakitai** [comment]
> #3 

Ah, yes, it looks better. I will check if it works tomorrow (probably it works), and merge it.

> README

Yeah, there is no "platform above"... I'm going to delete it.

> Renaming `arx::arx_std` to e.g. `arx::stdx`

I prefer `arx::stdx` because it's short and pretty (and doesn't lack the meaning) for me :)
I'm going to rename to `arx::stdx`.
